### DISCUSSION TITLE
WixDotNetCoreBootstrapperApplicationHost and Self Contained

### DISCUSSION THREAD

**Discussion by Steve0212a** [wixtoolset/issues#7400] [Questions]
I have a bundle with a custom UI that I just converted to .Net 6.  Currently, I am installing .Net 6 as a prereq for my custom WPF UI.

Questions:

1. Can I publish the custom UI as self-contained and then bundle all that up into the installer?  The goal would be that I would not need to install .Net 6 as a prereq.
2. If so, is there a way to do a "heat like" scan of the publish folder to create the "<Payload SourceFile=" entries?

Thanks

**Comment by BMurri** [comment]
Would publishing self-contained single-file work for you? There's only one binary provided by you that way that you need to include in your payload.

**Reply by rseanhall** [reply]
No, that wouldn't work. This is for a BootstrapperApplication so it has to be a DLL. An example of what they're asking for is in the wix4 repo.

**Reply by Steve0212a** [reply]
thx - I will look at the repo

**Reply by Steve0212a** [reply]
I searched the whole repository for "self-contained" and all I found was https://github.com/wixtoolset/wix4/blob/develop/src/ext/Bal/test/examples/examples.proj (other than a comment in another file).  

Is that what you are pointing me to?

**Reply by BMurri** [reply]
@rseanhall according to this question on stackoverflow, self-contained libraries are a thing that works (for at least the last 5 years): https://stackoverflow.com/questions/49141073/self-contained-net-core-library-missing-net-core-runtime-files

**Reply by Steve0212a** [reply]
@BMurri to be clear, I think @rseanhall is saying the single file will not work.  I tried to publish my UI Dll as a single file and got this error message (which makes sense):

`error NETSDK1099: Publishing to a single-file is only supported for executable applications`

**Reply by BMurri** [reply]
I hadn't tried it yet, but I'd seen reports like the one I shared a few times over the years, including recently

**Reply by rseanhall** [reply]
There is a difference between publishing self-contained and self-contained single file. An MBA can be published self-contained but not self-contained single file. Technically speaking, .NET does not support publishing libraries as self-contained at all (https://github.com/dotnet/sdk/issues/12043) even though it works. Publishing a library as self-contained single file is never going to work because .NET Core can't load managed code from .exe. The only reason that worked in .NET Framework was because of how it integrated with the OS.

**Comment by rseanhall** [comment]
There is an example of this in the v4 repo. You already found the [part](https://github.com/wixtoolset/wix4/blob/8c757c0f67f26f21c6bcbbfb81b7ea8b91c35fe4/src/ext/Bal/test/examples/examples.proj#L40) where the MBA is published as self-contained (note: not single file). In the [.wxs](https://github.com/wixtoolset/wix4/blob/8c757c0f67f26f21c6bcbbfb81b7ea8b91c35fe4/src/ext/Bal/test/examples/LatestCoreBundleSCD/SelfContainedBundle.wxs#L4), you can see where it adds `SelfContainedDeployment="yes"` to the `WixDotNetCoreBootstrapperApplicationHost` element, as well as the `PayloadGroupRef` to the files harvested by Heat. In the [.wixproj](https://github.com/wixtoolset/wix4/blob/8c757c0f67f26f21c6bcbbfb81b7ea8b91c35fe4/src/ext/Bal/test/examples/LatestCoreBundleSCD/LatestCoreBundleSCD.wixproj#L5) you can see how it set up Heat with the `HarvestDirectory` although you'll also need the HarvestDirectoryAdditionalOptions from the [.wixproj.props](https://github.com/wixtoolset/wix4/blob/8c757c0f67f26f21c6bcbbfb81b7ea8b91c35fe4/src/ext/Bal/test/examples/Directory.wixproj.props#L7).

**Reply by Steve0212a** [reply]
Thanks for your help on this - I finally got back to it and I was able to get it working.

Question - is there any way to put a status screen up while the bootstrapper is being extracted.  My whole installer is large at 380 MB.  Now, when I run my installer, it is about 12-15 seconds before the user sees anything.  Something like "extracting...." I think would be helpful for the user so they know something is running and don't run it a second time.

**Reply by barnson** [reply]
Bundle/@SplashScreenSourceFile

**Reply by Steve0212a** [reply]
Thanks - did not know that existed.  Just as an FYI for anyone else reading this, you need to call CloseSplashScreen() on the engine to close it...
